File: packages/python/src/telebot_py/types/media.py
"""Telegram media file types (photos, audio, video, documents)."""

from __future__ import annotations

import dataclasses

from telebot_py.types.base import TelegramObject


@dataclasses.dataclass(frozen=True, slots=True)
class PhotoSize(TelegramObject):
    """One size of a photo or a file / sticker thumbnail.

    Attributes:
        file_id: Identifier for this file, usable for downloading or reuse.
        file_unique_id: Unique, persistent identifier of this file.
        width: Photo width.
        height: Photo height.
        file_size: File size in bytes, when known.

    Telegram API: https://core.telegram.org/bots/api#photosize
    """

    file_id: str
    file_unique_id: str
    width: int
    height: int
    file_size: int | None = None


@dataclasses.dataclass(frozen=True, slots=True)
class Audio(TelegramObject):
    """An audio file to be treated as music by the Telegram clients.

    Attributes:
        file_id: Identifier for this file, usable for downloading or reuse.
        file_unique_id: Unique, persistent identifier of this file.
        duration: Duration of the audio in seconds as defined by sender.
        performer: Performer of the audio as defined by sender or tags.
        title: Title of the audio as defined by sender or tags.
        file_name: Original filename as defined by sender.
        mime_type: MIME type of the file as defined by sender.
        file_size: File size in bytes, when known.
        thumbnail: Thumbnail of the album cover the music file belongs to.

    Telegram API: https://core.telegram.org/bots/api#audio
    """

    file_id: str
    file_unique_id: str
    duration: int
    performer: str | None = None
    title: str | None = None
    file_name: str | None = None
    mime_type: str | None = None
    file_size: int | None = None
    thumbnail: PhotoSize | None = None


@dataclasses.dataclass(frozen=True, slots=True)
class Document(TelegramObject):
    """A general file (as opposed to photos, audio or video files).

    Attributes:
        file_id: Identifier for this file, usable for downloading or reuse.
        file_unique_id: Unique, persistent identifier of this file.
        thumbnail: Document thumbnail as defined by sender.
        file_name: Original filename as defined by sender.
        mime_type: MIME type of the file as defined by sender.
        file_size: File size in bytes, when known.

    Telegram API: https://core.telegram.org/bots/api#document
    """

    file_id: str
    file_unique_id: str
    thumbnail: PhotoSize | None = None
    file_name: str | None = None
    mime_type: str | None = None
    file_size: int | None = None


@dataclasses.dataclass(frozen=True, slots=True)
class VideoQuality(TelegramObject):
    """A video file of one specific quality.

    Attributes:
        file_id: Identifier for this file, usable for downloading or reuse.
        file_unique_id: Unique, persistent identifier of this file.
        width: Video width.
        height: Video height.
        codec: Codec used to encode the video, e.g. "h264", "h265" or "av01".
        file_size: File size in bytes, when known.

    Telegram API: https://core.telegram.org/bots/api#videoquality
    """

    file_id: str
    file_unique_id: str
    width: int
    height: int
    codec: str
    file_size: int | None = None


@dataclasses.dataclass(frozen=True, slots=True)
class Video(TelegramObject):
    """A video file.

    Attributes:
        file_id: Identifier for this file, usable for downloading or reuse.
        file_unique_id: Unique, persistent identifier of this file.
        width: Video width as defined by sender.
        height: Video height as defined by sender.
        duration: Duration of the video in seconds as defined by sender.
        thumbnail: Video thumbnail.
        cover: Available sizes of the cover of the video.
        start_timestamp: Timestamp in seconds from which the video plays in the
            message.
        qualities: List of available qualities of the video.
        file_name: Original filename as defined by sender.
        mime_type: MIME type of the file as defined by sender.
        file_size: File size in bytes, when known.

    Telegram API: https://core.telegram.org/bots/api#video
    """

    file_id: str
    file_unique_id: str
    width: int
    height: int
    duration: int
    thumbnail: PhotoSize | None = None
    cover: list[PhotoSize] | None = None
    start_timestamp: int | None = None
    qualities: list[VideoQuality] | None = None
    file_name: str | None = None
    mime_type: str | None = None
    file_size: int | None = None


@dataclasses.dataclass(frozen=True, slots=True)
class Animation(TelegramObject):
    """An animation file (GIF or H.264/MPEG-4 AVC video without sound).

    Attributes:
        file_id: Identifier for this file, usable for downloading or reuse.
        file_unique_id: Unique, persistent identifier of this file.
        width: Video width as defined by sender.
        height: Video height as defined by sender.
        duration: Duration of the video in seconds as defined by sender.
        thumbnail: Animation thumbnail.
        file_name: Original animation filename as defined by sender.
        mime_type: MIME type of the file as defined by sender.
        file_size: File size in bytes, when known.

    Telegram API: https://core.telegram.org/bots/api#animation
    """

    file_id: str
    file_unique_id: str
    width: int
    height: int
    duration: int
    thumbnail: PhotoSize | None = None
    file_name: str | None = None
    mime_type: str | None = None
    file_size: int | None = None


@dataclasses.dataclass(frozen=True, slots=True)
class Voice(TelegramObject):
    """A voice note.

    Attributes:
        file_id: Identifier for this file, usable for downloading or reuse.
        file_unique_id: Unique, persistent identifier of this file.
        duration: Duration of the audio in seconds as defined by sender.
        mime_type: MIME type of the audio as defined by sender.
        file_size: File size in bytes, when known.

    Telegram API: https://core.telegram.org/bots/api#voice
    """

    file_id: str
    file_unique_id: str
    duration: int
    mime_type: str | None = None
    file_size: int | None = None


@dataclasses.dataclass(frozen=True, slots=True)
class VideoNote(TelegramObject):
    """A video message (round video).

    Attributes:
        file_id: Identifier for this file, usable for downloading or reuse.
        file_unique_id: Unique, persistent identifier of this file.
        length: Video width and height (diameter) as defined by sender.
        duration: Duration of the video in seconds as defined by sender.
        thumbnail: Video thumbnail.
        file_size: File size in bytes, when known.

    Telegram API: https://core.telegram.org/bots/api#videonote
    """

    file_id: str
    file_unique_id: str
    length: int
    duration: int
    thumbnail: PhotoSize | None = None
    file_size: int | None = None


@dataclasses.dataclass(frozen=True, slots=True)
class LivePhoto(TelegramObject):
    """A live photo: a still photo paired with the video it animates into.

    Attributes:
        file_id: Identifier for the video file, usable for downloading or
            reuse.
        file_unique_id: Unique, persistent identifier of the video file.
        width: Video width as defined by sender.
        height: Video height as defined by sender.
        duration: Duration of the video in seconds as defined by sender.
        photo: Available sizes of the corresponding static photo.
        mime_type: MIME type of the video file as defined by sender.
        file_size: File size of the video in bytes, when known.

    Telegram API: https://core.telegram.org/bots/api#livephoto
    """

    file_id: str
    file_unique_id: str
    width: int
    height: int
    duration: int
    photo: list[PhotoSize] | None = None
    mime_type: str | None = None
    file_size: int | None = None

File: packages/node/src/client/types/messages/media.ts
import type { ParseMode } from "../../constants.js";
import type { MessageEntity } from "./core.js";
import type { InputFile } from "../../../utils/http.js";

export type { InputFile };

/**
 * Represents one size of a photo or a file / sticker thumbnail.
 *
 * @see {@link https://core.telegram.org/bots/api#photosize Telegram Bot API: PhotoSize}
 */
export interface PhotoSize {
  /** Identifier for this file, which can be used to download or reuse the file. */
  file_id: string;
  /** Unique identifier for this file, which is supposed to be the same over time and for different bots. */
  file_unique_id: string;
  /** Photo width. */
  width: number;
  /** Photo height. */
  height: number;
  /** File size in bytes. */
  file_size?: number;
}

/**
 * Represents an audio file to be treated as music by the Telegram clients.
 *
 * @see {@link https://core.telegram.org/bots/api#audio Telegram Bot API: Audio}
 */
export interface Audio {
  /** Identifier for this file, which can be used to download or reuse the file. */
  file_id: string;
  /** Unique identifier for this file, which is supposed to be the same over time and for different bots. */
  file_unique_id: string;
  /** Duration of the audio in seconds as defined by sender. */
  duration: number;
  /** Performer of the audio as defined by sender or by audio tags. */
  performer?: string;
  /** Title of the audio as defined by sender or by audio tags. */
  title?: string;
  /** Original filename as defined by sender. */
  file_name?: string;
  /** MIME type of the file as defined by sender. */
  mime_type?: string;
  /** File size in bytes. */
  file_size?: number;
  /** Thumbnail of the album cover to which the music file belongs. */
  thumbnail?: PhotoSize;
}

/**
 * Represents a general file (as opposed for photos or audio files).
 *
 * @see {@link https://core.telegram.org/bots/api#document Telegram Bot API: Document}
 */
export interface Document {
  /** Identifier for this file, which can be used to download or reuse the file. */
  file_id: string;
  /** Unique identifier for this file, which is supposed to be the same over time and for different bots. */
  file_unique_id: string;
  /** Document thumbnail as defined by sender. */
  thumbnail?: PhotoSize;
  /** Original filename as defined by sender. */
  file_name?: string;
  /** MIME type of the file as defined by sender. */
  mime_type?: string;
  /** File size in bytes. */
  file_size?: number;
}

/**
 * Represents a video file of a specific quality.
 *
 * @see {@link https://core.telegram.org/bots/api#videoquality Telegram Bot API: VideoQuality}
 */
export interface VideoQuality {
  /** Identifier for this file, which can be used to download or reuse the file. */
  file_id: string;
  /** Unique identifier for this file, which is supposed to be the same over time and for different bots. */
  file_unique_id: string;
  /** Video width. */
  width: number;
  /** Video height. */
  height: number;
  /** Codec that was used to encode the video, for example, "h264", "h265", or "av01". */
  codec: string;
  /** File size in bytes. */
  file_size?: number;
}

/**
 * Represents a video file.
 *
 * @see {@link https://core.telegram.org/bots/api#video Telegram Bot API: Video}
 */
export interface Video {
  /** Identifier for this file, which can be used to download or reuse the file. */
  file_id: string;
  /** Unique identifier for this file, which is supposed to be the same over time and for different bots. */
  file_unique_id: string;
  /** Video width as defined by sender. */
  width: number;
  /** Video height as defined by sender. */
  height: number;
  /** Duration of the video in seconds as defined by sender. */
  duration: number;
  /** Video thumbnail. */
  thumbnail?: PhotoSize;
  /** Available sizes of the cover of the video in the message. */
  cover?: PhotoSize[];
  /** Timestamp in seconds from which the video will play in the message. */
  start_timestamp?: number;
  /** List of available qualities of the video. */
  qualities?: VideoQuality[];
  /** Original filename as defined by sender. */
  file_name?: string;
  /** MIME type of the file as defined by sender. */
  mime_type?: string;
  /** File size in bytes. */
  file_size?: number;
}

/**
 * Represents an animation file (GIF or H.264/MPEG-4 AVC video without sound).
 *
 * @see {@link https://core.telegram.org/bots/api#animation Telegram Bot API: Animation}
 */
export interface Animation {
  /** Identifier for this file, which can be used to download or reuse the file. */
  file_id: string;
  /** Unique identifier for this file, which is supposed to be the same over time and for different bots. */
  file_unique_id: string;
  /** Video width as defined by sender. */
  width: number;
  /** Video height as defined by sender. */
  height: number;
  /** Duration of the video in seconds as defined by sender. */
  duration: number;
  /** Animation thumbnail. */
  thumbnail?: PhotoSize;
  /** Original animation filename as defined by sender. */
  file_name?: string;
  /** MIME type of the file as defined by sender. */
  mime_type?: string;
  /** File size in bytes. */
  file_size?: number;
}

/**
 * Represents a voice note.
 *
 * @see {@link https://core.telegram.org/bots/api#voice Telegram Bot API: Voice}
 */
export interface Voice {
  /** Identifier for this file, which can be used to download or reuse the file. */
  file_id: string;
  /** Unique identifier for this file, which is supposed to be the same over time and for different bots. */
  file_unique_id: string;
  /** Duration of the audio in seconds as defined by sender. */
  duration: number;
  /** MIME type of the audio as defined by sender. */
  mime_type?: string;
  /** File size in bytes. */
  file_size?: number;
}

/**
 * Represents a video message (round video).
 *
 * @see {@link https://core.telegram.org/bots/api#videonote Telegram Bot API: VideoNote}
 */
export interface VideoNote {
  /** Identifier for this file, which can be used to download or reuse the file. */
  file_id: string;
  /** Unique identifier for this file, which is supposed to be the same over time and for different bots. */
  file_unique_id: string;
  /** Video width and height (diameter of the video message) as defined by sender. */
  length: number;
  /** Duration of the video in seconds as defined by sender. */
  duration: number;
  /** Video thumbnail. */
  thumbnail?: PhotoSize;
  /** File size in bytes. */
  file_size?: number;
}

/**
 * Represents a Live Photo message object (Bot API 10.3+).
 *
 * @see {@link https://core.telegram.org/bots/api#livephoto Telegram Bot API: LivePhoto}
 */
export interface LivePhoto {
  /** Available sizes of the corresponding static photo. */
  photo?: PhotoSize[];
  /** Identifier for the video file which can be used to download or reuse the file. */
  file_id: string;
  /** Unique identifier for the video file which is supposed to be the same over time and for different bots. */
  file_unique_id: string;
  /** Video width as defined by the sender. */
  width: number;
  /** Video height as defined by the sender. */
  height: number;
  /** Duration of the video in seconds as defined by the sender. */
  duration: number;
  /** MIME type of the file as defined by the sender. */
  mime_type?: string;
  /** File size in bytes. */
  file_size?: number;
}

/**
 * Represents a photo to be sent as part of a media group or edited media.
 *
 * @see {@link https://core.telegram.org/bots/api#inputmediaphoto Telegram Bot API: InputMediaPhoto}
 */
export interface InputMediaPhoto {
  /** Type of the result, must be 'photo'. */
  type: "photo";
  /** File to send. Pass a file_id to send a file that exists on the Telegram servers, an HTTP URL, or upload using InputFile. */
  media: string | InputFile;
  /** Caption of the photo to be sent, 0-1024 characters after entities parsing. */
  caption?: string;
  /** Mode for parsing entities in the photo caption. */
  parse_mode?: ParseMode;
  /** List of special entities that appear in the caption. */
  caption_entities?: MessageEntity[];
  /** Pass True, if the caption must be shown above the message media. */
  show_caption_above_media?: boolean;
  /** Pass True, if the photo needs to be covered with a spoiler animation. */
  has_spoiler?: boolean;
}

/**
 * Represents a video to be sent as part of a media group or edited media.
 *
 * @see {@link https://core.telegram.org/bots/api#inputmediavideo Telegram Bot API: InputMediaVideo}
 */
export interface InputMediaVideo {
  /** Type of the result, must be 'video'. */
  type: "video";
  /** File to send. Pass a file_id to send a file that exists on the Telegram servers, an HTTP URL, or upload using InputFile. */
  media: string | InputFile;
  /** Thumbnail of the file sent. */
  thumbnail?: string | InputFile;
  /** Cover for the video in the message. Pass a file_id, an HTTP URL, or upload using InputFile. */
  cover?: string | InputFile;
  /** Start timestamp for the video in the message. */
  start_timestamp?: number;
  /** Caption of the video to be sent, 0-1024 characters after entities parsing. */
  caption?: string;
  /** Mode for parsing entities in the video caption. */
  parse_mode?: ParseMode;
  /** List of special entities that appear in the caption. */
  caption_entities?: MessageEntity[];
  /** Pass True, if the caption must be shown above the message media. */
  show_caption_above_media?: boolean;
  /** Video width. */
  width?: number;
  /** Video height. */
  height?: number;
  /** Video duration in seconds. */
  duration?: number;
  /** Pass True, if the uploaded video is suitable for streaming. */
  supports_streaming?: boolean;
  /** Pass True, if the video needs to be covered with a spoiler animation. */
  has_spoiler?: boolean;
}

/**
 * Represents an animation file to be sent as part of a media group or edited media.
 *
 * @see {@link https://core.telegram.org/bots/api#inputmediaanimation Telegram Bot API: InputMediaAnimation}
 */
export interface InputMediaAnimation {
  /** Type of the result, must be 'animation'. */
  type: "animation";
  /** File to send. Pass a file_id to send a file that exists on the Telegram servers, an HTTP URL, or upload using InputFile. */
  media: string | InputFile;
  /** Thumbnail of the file sent. */
  thumbnail?: string | InputFile;
  /** Caption of the animation to be sent, 0-1024 characters after entities parsing. */
  caption?: string;
  /** Mode for parsing entities in the animation caption. */
  parse_mode?: ParseMode;
  /** List of special entities that appear in the caption. */
  caption_entities?: MessageEntity[];
  /** Pass True, if the caption must be shown above the message media. */
  show_caption_above_media?: boolean;
  /** Animation width. */
  width?: number;
  /** Animation height. */
  height?: number;
  /** Animation duration in seconds. */
  duration?: number;
  /** Pass True, if the animation needs to be covered with a spoiler animation. */
  has_spoiler?: boolean;
}

/**
 * Represents an audio file to be treated as music to be sent as part of a media group or edited media.
 *
 * @see {@link https://core.telegram.org/bots/api#inputmediaaudio Telegram Bot API: InputMediaAudio}
 */
export interface InputMediaAudio {
  /** Type of the result, must be 'audio'. */
  type: "audio";
  /** File to send. Pass a file_id to send a file that exists on the Telegram servers, an HTTP URL, or upload using InputFile. */
  media: string | InputFile;
  /** Thumbnail of the file sent. */
  thumbnail?: string | InputFile;
  /** Caption of the audio to be sent, 0-1024 characters after entities parsing. */
  caption?: string;
  /** Mode for parsing entities in the audio caption. */
  parse_mode?: ParseMode;
  /** List of special entities that appear in the caption. */
  caption_entities?: MessageEntity[];
  /** Duration of the audio in seconds. */
  duration?: number;
  /** Performer of the audio. */
  performer?: string;
  /** Title of the audio. */
  title?: string;
}

/**
 * Represents a general file to be sent as part of a media group or edited media.
 *
 * @see {@link https://core.telegram.org/bots/api#inputmediadocument Telegram Bot API: InputMediaDocument}
 */
export interface InputMediaDocument {
  /** Type of the result, must be 'document'. */
  type: "document";
  /** File to send. Pass a file_id to send a file that exists on the Telegram servers, an HTTP URL, or upload using InputFile. */
  media: string | InputFile;
  /** Thumbnail of the file sent. */
  thumbnail?: string | InputFile;
  /** Caption of the document to be sent, 0-1024 characters after entities parsing. */
  caption?: string;
  /** Mode for parsing entities in the document caption. */
  parse_mode?: ParseMode;
  /** List of special entities that appear in the caption. */
  caption_entities?: MessageEntity[];
  /** Disables automatic server-side content type detection for files. */
  disable_content_type_detection?: boolean;
}

/**
 * Union of all input media types that can be sent in a media group or editMessageMedia.
 *
 * @see {@link https://core.telegram.org/bots/api#inputmedia Telegram Bot API: InputMedia}
 */
export type InputMedia =
  InputMediaPhoto | InputMediaVideo | InputMediaAnimation | InputMediaAudio | InputMediaDocument;

/**
 * The paid media to send is a photo.
 *
 * @see {@link https://core.telegram.org/bots/api#inputpaidmediaphoto Telegram Bot API: InputPaidMediaPhoto}
 */
export interface InputPaidMediaPhoto {
  /** Type of the media, must be 'photo'. */
  type: "photo";
  /** File to send. Pass a file_id to send a file that exists on the Telegram servers, or an HTTP URL. */
  media: string;
}

/**
 * The paid media to send is a video.
 *
 * @see {@link https://core.telegram.org/bots/api#inputpaidmediavideo Telegram Bot API: InputPaidMediaVideo}
 */
export interface InputPaidMediaVideo {
  /** Type of the media, must be 'video'. */
  type: "video";
  /** File to send. Pass a file_id to send a file that exists on the Telegram servers, or an HTTP URL. */
  media: string;
  /** Thumbnail of the file sent. */
  thumbnail?: string;
  /** Cover for the video in the message. */
  cover?: string;
  /** Start timestamp for the video in the message. */
  start_timestamp?: number;
  /** Video width. */
  width?: number;
  /** Video height. */
  height?: number;
  /** Video duration in seconds. */
  duration?: number;
  /** Pass True, if the uploaded video is suitable for streaming. */
  supports_streaming?: boolean;
}

/**
 * The paid media to send is a live photo.
 *
 * @see {@link https://core.telegram.org/bots/api#inputpaidmedialivephoto Telegram Bot API: InputPaidMediaLivePhoto}
 */
export interface InputPaidMediaLivePhoto {
  /** Type of the media, must be 'live_photo'. */
  type: "live_photo";
  /** File to send. Pass a file_id to send a file that exists on the Telegram servers, or an HTTP URL. */
  media: string;
  /** Photo to send. Pass a file_id to send a file that exists on the Telegram servers, or an HTTP URL. */
  photo: string;
}

/**
 * Describes the paid media to be sent.
 *
 * @see {@link https://core.telegram.org/bots/api#inputpaidmedia Telegram Bot API: InputPaidMedia}
 */
export type InputPaidMedia = InputPaidMediaPhoto | InputPaidMediaVideo | InputPaidMediaLivePhoto;
